python/tank/bootstrap/manager.py

```diff
@@ -221,49 +247,160 @@ def set_progress_callback(self, callback):
+        an :class:`sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+
         If entity is None, the method will bootstrap into the site
         config. This method will attempt to resolve the config according
         to business logic set in the associated resolver class and based
         on this launch a configuration. This may involve downloading new
         apps from the toolkit app store and installing files on disk.
 
+        3 callback functions can be provided.
+
+        A callback function that reports back on the bootstrap progress
+        with the following signature::
+
+            progress_callback(step_number, message, current_index, maximum_index)
+
+        where:
+        - ``step_number`` is the current progress step number,
+                          from 1 to ``ToolkitManager.MAX_PROGRESS_STEP_NUMBER``.
+        - ``message`` is the progress message to report.
+        - ``current_index`` is an optional current item number being looped over.
+        - ``maximum_index`` is an optional maximum item number being looped over.
+
+        A callback function that handles cleanup after successful completion of the bootstrap
+        with the following signature::
+
+            completed_callback(engine)
+
+        where:
+        - ``engine``is the launched :class:`sgtk.platform.Engine` instance.
+
+        A callback function that handles cleanup after failed completion of the bootstrap
+        with the following signature::
+
+            failed_callback(step, exception)
+
+        where:
+        - ``step``is the bootstrap step ("sgtk" or "engine") that raised the exception.
+        - ``exception``is the python exception raised while bootstrapping.
+
         Please note that the API version of the tk instance that hosts
         the engine may not be the same as the API version that was
         executed during the bootstrap.
 
-        :param entity: Shotgun entity to launch engine for
-        :type entity: Dictionary with keys type and id
+        :param engine_name: Name of engine to launch (e.g. ``tk-nuke``).
+        :param entity: Shotgun entity to launch engine for.
+        :param progress_callback: Callback function that reports back on the toolkit and engine bootstrap progress.
+        :param completed_callback: Callback function that handles cleanup after successful completion of the bootstrap.
+        :param failed_callback: Callback function that handles cleanup after failed completion of the bootstrap.
+        :returns: :class:`sgtk.platform.Engine` instance when ``async_bootstrap` is false; otherwise None.
+        """
+
+        log.info("Bootstrapping engine %s for entity %s." % (engine_name, entity))
+
+        bootstrap_asynchronously = self.async_bootstrap
+
+        if bootstrap_asynchronously:
+            try:
+                import threading
```

Comment:
`import threading` raises an `ImportError` exception when Qt is not available, and the manager can fall back on synchronous bootstrapping rather than calling the threading asynchronous bootstrapper.